src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -2904,116 +2816,135 @@ private static bool IsReallyAType(TypeSymbol type)
-                foreach (var value in list)
+                foreach (var candidate in initialCandidates)
                 {
-                    if (!_valueComparer.Equals(type, value))
+                    if (bound.Equals(candidate, comparison))
                     {
-                        builder.Add(value);
+                        continue;
+                    }
+                    TypeSymbolWithAnnotations source;
+                    TypeSymbolWithAnnotations destination;
+                    if (variance == VarianceKind.Out)
+                    {
+                        source = bound;
+                        destination = candidate;
+                    }
+                    else
+                    {
+                        source = candidate;
+                        destination = bound;
+                    }
+                    if (!ImplicitConversionExists(source, destination, ref useSiteDiagnostics, conversions.WithNullability(false)))
+                    {
+                        candidates.Remove(candidate);
+                    }
+                    else if (bound.Equals(candidate, TypeCompareKind.IgnoreDynamicAndTupleNames))
+                    {
+                        // SPEC: 4.7 The Dynamic Type
+                        //       Type inference (7.5.2) will prefer dynamic over object if both are candidates.
+                        //
+                        // This rule doesn't have to be implemented explicitly due to special handling of 
+                        // conversions from dynamic in ImplicitConversionExists helper.
+                        // 
+                        MergeAndReplaceIfStillCandidate(candidates, candidate, bound, variance, conversions, out bool hadMismatch);
+                        hadNullabilityMismatch |= hadMismatch;
                     }
-                }
-                if (builder.Count == 0)
-                {
-                    _dictionary.Remove(type);
-                }
-                else
-                {
-                    _dictionary[type] = builder;
                 }
             }
         }
 
-        private static TypeCompareKind GetComparison(bool includeNullability)
+        private static void MergeAndReplaceIfStillCandidate(
+            Dictionary<TypeSymbolWithAnnotations, TypeSymbolWithAnnotations> candidates,
+            TypeSymbolWithAnnotations old,
+            TypeSymbolWithAnnotations @new,
+            VarianceKind variance,
+            ConversionsBase conversions,
+            out bool hadNullabilityMismatch)
         {
-            return includeNullability?
-                TypeCompareKind.IgnoreDynamicAndTupleNames | TypeCompareKind.CompareNullableModifiersForReferenceTypes :
-                TypeCompareKind.IgnoreDynamicAndTupleNames;
+            hadNullabilityMismatch = false;
+
+            // We make an exception when @new is dynamic, for backwards compatibility 
+            if (@new.IsDynamic())
```

Comment:
I don't understand this comment or the code that follows it. Can you give an example?

Reply:
This is code that existed before the original nullable change to `MethodTypeInferrer`. See [https://github.com/dotnet/roslyn/blob/dev15.8.x/src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs#L2821](https://github.com/dotnet/roslyn/blob/dev15.8.x/src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs#L2821).

---
In reply to: [225363137](https://github.com/dotnet/roslyn/pull/29727#discussion_r225363137) [](ancestors = 225363137)